packages/react-next/src/components/Checkbox/Checkbox.base.tsx

```diff
@@ -1,152 +1,60 @@
-          <label className={classNames.label} htmlFor={id}>
-            <div className={classNames.checkbox} data-ktp-target={keytipAttributes['data-ktp-target']}>
-              <Icon iconName="CheckMark" {...checkmarkIconProps} className={classNames.checkmark} />
-            </div>
-            {(props.onRenderLabel || onRenderLabel)(props, onRenderLabel)}
-          </label>
-        </div>
-      )}
-    </KeytipData>
-  );
-});
-
-CheckboxBase.displayName = 'CheckboxBase';
-
-function useDebugWarning(props: ICheckboxProps) {
-  if (process.env.NODE_ENV !== 'production') {
-    // This is a build-time conditional that will be constant at runtime
-    // tslint:disable-next-line:react-hooks-nesting
-    React.useEffect(() => {
-      warnMutuallyExclusive('Checkbox', props, {
-        checked: 'defaultChecked',
-        indeterminate: 'defaultIndeterminate',
-      });
-    }, []);
-  }
-}
-
-function useComponentRef(
-  props: ICheckboxProps,
-  isChecked: boolean | undefined,
-  isIndeterminate: boolean | undefined,
-  checkBox: React.RefObject<HTMLInputElement>,
-) {
-  React.useImperativeHandle(
-    props.componentRef,
-    () => ({
-      get checked() {
-        return !!isChecked;
-      },
-      get indeterminate() {
-        return !!isIndeterminate;
-      },
-      focus() {
-        if (checkBox.current) {
-          checkBox.current.focus();
-        }
-      },
-    }),
-    [isChecked, isIndeterminate],
-  );
-}
+import { useCheckbox } from './useCheckbox';
+
+const defaultSlots: Omit<ICheckboxSlots, 'root'> = {
+  input: 'input',
+  checkmark: Icon,
+  container: 'label',
+  text: 'span',
+  checkbox: 'div',
+};
+
+export const CheckboxBase = compose<'div', ICheckboxProps, ICheckboxProps, {}, {}>(
+  (props, forwardedRef, composeOptions) => {
+    const { slotProps, slots, state } = useCheckbox(props, composeOptions, forwardedRef);
+    const { disabled, keytipProps, label } = state;
+
+    const onRenderLabel = (): JSX.Element | null => {
+      return label ? <slots.text {...slotProps.text}>{label}</slots.text> : null;
+    };
+
+    return (
+      <KeytipData keytipProps={keytipProps} disabled={disabled}>
+        {// tslint:disable-next-line:no-any
+        (keytipAttributes: any): JSX.Element => (
+          <slots.root {...slotProps.root}>
+            <slots.input
+              {...slotProps.input}
+              data-ktp-execute-target={keytipAttributes['data-ktp-execute-target']}
+              aria-describedby={mergeAriaAttributeValues(
+                slotProps.input['aria-describedby'],
+                keytipAttributes['aria-describedby'],
+              )}
+            />
+            <slots.container {...slotProps.container}>
+              <slots.checkbox {...slotProps.checkbox} data-ktp-target={keytipAttributes['data-ktp-target']}>
+                <slots.checkmark {...slotProps.checkmark} />
+              </slots.checkbox>
+              {(props.onRenderLabel || onRenderLabel)(props, onRenderLabel)}
```

Comment:
shouldn't compse and shorthands allow us to stop using custom `onRender` props by using children callback instead? @dzearing

Reply:
Yup! but if we remove this, it's a breaking change so i am leaving it here. i think the right thing do to is to deprecate and promote using slot

Reply:
I think we need to break. we should not use the new AND the old patterns, and we should instead rely on the compat layer imo to optimize perf of the new one.

Reply:
if you render a label slot, there is a clear upgrade path and it works as expected.